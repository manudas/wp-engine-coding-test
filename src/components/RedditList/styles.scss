.reddit-list {
    &_posts {
        margin: 5px;
    }
    &_header {
        margin: 5px;

        display: flex;
        flex-direction: row;
        justify-content: space-between;
        
        &-page {
            margin: 0 5px;
        }

        &-button {
            transition: all 0.3s ease-out;
            
            &.disabled {
                pointer-events: none;
                opacity: 0.7;
            }
        }
    }
}