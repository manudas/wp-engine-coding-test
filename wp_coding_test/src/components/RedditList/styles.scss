@import '../globals';

.reddit-list {
    &__posts {
        margin: 5px;

        display: flex;
        flex-wrap: wrap;
        justify-content: center;
    }
    &__header {
        margin: 5px;

        display: flex;
        flex-direction: row;
        justify-content: space-between;

        align-items: center;

        background-color: $semiTransparentOrange;

        padding: 5px;
        border-radius: 3px;
    }
}
