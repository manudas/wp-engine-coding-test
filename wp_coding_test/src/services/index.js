import {
    endpoint,
} from '../components/constants';

const getPostsFromPayload = ({
    data: {
        children: postsList
    }
}) => {

    return postsList.map(({data}) => data).sort((a, b) => Number(a.created_utc) - Number(b.created_utc));
};

export const fetchSubredditPosts = async (subreddit) => {

    const url = `${endpoint}${subreddit}.json`;
    try {
        const response = await fetch(encodeURI(url));
        if (response.ok) {
            const payload = await response.json();
            const posts = getPostsFromPayload(payload);
            return posts;
        }
    } catch(err) {
        // external is _not_ reachable
        console.error('Network error, url not reachable');
    };

    return [];
};