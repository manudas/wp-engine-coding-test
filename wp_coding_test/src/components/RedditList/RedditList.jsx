import React, {
    useState,
    useEffect,
} from 'react';

import { fetchSubredditPosts } from '../../services';

import RedditPost from '../RedditPost';
import DebouncedInput from '../DebouncedInput';

import {
    postsPerPage,
    defaultReddit,
} from '../constants';

import './styles.scss';

const RedditList = () => {
    const [ page, setPage ] = useState(0);
    const [ posts, setPosts ] = useState([]);

    useEffect(() => {
        onChangeInputHandler(defaultReddit);
    }, []);

    const onChangeInputHandler = async (subReddit) => {
        const posts = await fetchSubredditPosts(subReddit);
        setPosts(posts);
        setPage(0);
    }

    const firstPaginatedElement = (postsPerPage * page);
    const lastPaginatedElement = (postsPerPage * page) + postsPerPage;
    const maxPage = Math.floor(posts.length / postsPerPage);

    return (
        <>
            <header
                className="reddit-list_header"
            >
                <DebouncedInput
                    onChangeHandler={onChangeInputHandler}
                />
                <div>
                    <button
                        className={`reddit-list_header-button ${page === 0 ? 'disabled' : ''}`}
                        onClick={() => setPage(page - 1)}
                    >
                        Previous
                    </button>
                    <span
                        className="reddit-list_header-page"
                    >
                        Page {page + 1} out of {maxPage + 1}
                    </span>
                    <button
                        className={`reddit-list_header-button ${page === maxPage ? 'disabled' : ''}`}
                        onClick={() => setPage(page + 1)}
                    >
                        Next
                    </button>
                </div>
            </header>
            <div
                className="reddit-list_posts"
            >
                {
                    posts.slice(firstPaginatedElement, lastPaginatedElement).map(({
                        title,
                        thumbnail,
                        selftext,
                        id,
                        url,
                    }) =>
                        <RedditPost
                            key={id}
                            title={title}
                            thumbnail={thumbnail}
                            text={selftext}
                            url={url}
                        />
                    )
                }
            </div>
        </>
    );
};

export default RedditList;
