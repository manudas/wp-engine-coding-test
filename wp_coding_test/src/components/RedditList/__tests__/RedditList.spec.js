
import {
    render,
    waitFor,
    fireEvent,
} from '@testing-library/react';
import RedditList from '..';

import * as services from '../../../services'
import {
    goodDataResponse,
} from '../../../__data__';

describe(('Test suite for RedditList component'), () => {
    let renderer;

    it('should render an empty List and hit the API', async () => {
        jest.spyOn(services, 'fetchSubredditPosts').mockImplementation(() => []);

        renderer = render(<RedditList />);
        const { container } = renderer;
        expect(container).toMatchSnapshot();

        // wait for effects to have been run
        await waitFor(() => expect(services.fetchSubredditPosts).toHaveBeenCalledTimes(1));
        expect(services.fetchSubredditPosts).toHaveBeenCalledWith('reactjs');

        const expected = expect.arrayContaining([]);
        expect(services.fetchSubredditPosts).toEqual(expected);
    });

    it('should render a maximum of 10 post at at time', async () => {
        /*
         * Rather than mocking the service, we'll mock here
         * the call to window.fech. This way we'll have our
         * services run, so we are runing more of a kind of
         * a integration test here
         */
        jest.spyOn(window, 'fetch')
        window.fetch.mockResolvedValueOnce(goodDataResponse);

        renderer = render(<RedditList />);
        const { getAllByText } = renderer;

        // wait for effects to have been run
        // there's one link redirecting to the
        // post by subredit post drawn on screen
        expect((await waitFor(() => getAllByText(/go to post/i))).length).toBe(10);
    });

    it('should allow pagination when available', async () => {
        /*
         * We'll mock global fetch here too
         */
        jest.spyOn(window, 'fetch')
        window.fetch.mockResolvedValueOnce(goodDataResponse);

        renderer = render(<RedditList />);
        const {
            getAllByText,
            getByText,
        } = renderer;

        const previousButton = getByText(/previous/i);
        const nextButton = getByText(/next/i);

        // wait for effects to have been run
        // before checking the state of the buttons
        let goToPostLinks, nextPagePostLinks;
        await waitFor(() => goToPostLinks = nextPagePostLinks = getAllByText(/go to post/i));

        // first page: previous button should be disabled, next shouldn't
        expect(previousButton).toHaveClass('disabled');
        expect(nextButton).not.toHaveClass('disabled');

        fireEvent.click(nextButton); // page 2
        await waitFor(() => {
            // expect(goToPostLinks).not.toEqual(nextPagePostLinks); // testing expectation
            nextPagePostLinks = getAllByText(/go to post/i);
            expect(goToPostLinks).not.toEqual(nextPagePostLinks);
        });

        goToPostLinks = nextPagePostLinks;
        fireEvent.click(nextButton); // third page
        await waitFor(() => {
            nextPagePostLinks = getAllByText(/go to post/i);
            expect(goToPostLinks).not.toEqual(nextPagePostLinks);
        });

        /*
         * In our mocking data set we have
         * 27 posts. Being in the 3th page
         * we can only have 7 elements left
         * on the screen at this moment
         */
        expect(nextPagePostLinks.length).toBe(7);

        // last page: previous button should not be disabled, next should
        expect(previousButton).not.toHaveClass('disabled');
        expect(nextButton).toHaveClass('disabled');
    });

    afterEach(() => {
        jest.restoreAllMocks(); // clears mocked implementations
        jest.clearAllMocks();
    });
});
