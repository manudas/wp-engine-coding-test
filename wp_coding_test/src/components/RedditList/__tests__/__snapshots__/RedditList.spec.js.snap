// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite for RedditList component should render an empty List and hit the API 1`] = `
<div>
  <header
    class="reddit-list__header"
  >
    <label
      class="debounced-input__label"
      for="subreddit"
    >
      Please enter a subreddit
      <input
        class="debounced-input__input"
        data-testid="debouncedInput"
        id="subreddit"
        value="reactjs"
      />
    </label>
    <nav>
      <button
        class="paginator__button disabled"
      >
        Previous
      </button>
      <span
        class="paginator__page"
      >
        Page 
        1
         out of 
        1
      </span>
      <button
        class="paginator__button disabled"
      >
        Next
      </button>
    </nav>
  </header>
  <div
    class="reddit-list__posts"
  />
</div>
`;
